x/gamm/keeper/estimate_swap.go

```diff
@@ -0,0 +1,164 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v12/x/gamm/types"
+)
+
+// EstimateMultihopSwapExactAmountIn iterates `EstimateSwapExactAmountIn` and returns
+// the total token out given route for multihop.
+func (k Keeper) EstimateMultihopSwapExactAmountIn(
```

Comment:
There seems to be a lot of code copied over from non-estimate versions of this method. I'm wondering if it's possible to generalize this somehow so that we are not repeating so much boilerplate.

Reply:
It ranges over the same for loop for the multihops, the only difference is that it would call different internal methods, `swapExactAmountOut` for the actual swap and`estimateSwapExactAmountOut` for the estimated multihops.

This could be solved by passing in the internal func we want to use as a function parameter, but I think such refactor could be done in a separate PR

Reply:
https://github.com/osmosis-labs/osmosis/issues/3386